### PR TITLE
Create symlinks for symlink entries in tar when extracting

### LINKED ISSUE #2829: Failure sending layers to docker daemon
**Environment**:

- *Jib version: 2.6.0
- *Build tool: Gradle 6.3
- *OS: Centos 8
- *Docker: 19.03.13

**Description of the issue**:
We are building multiple images inside our pipeline job (Image: Centos 8). For that we are using the following command:
```bash
src/java/gradlew --gradle-user-home .gradle --info --stacktrace --build-cache --no-daemon --console=plain --parallel -p src/java jibDockerBuild -x test  -Djib.console=plain
```
Today we transferred our last service building its image to jib
The successfully build layers cannot be uploaded to the docker daemon. For all other services inside the job it is working.
```bash
Building dependencies layer...
Building project dependencies layer...
Building resources layer...
Building extra files layer...
Building classes layer...
Container program arguments set to [/bin/sh, -c, ./start.sh] (inherited from base image)
Loading to Docker daemon...
```
Out of 10 retries, one succeeded to upload to the docker daemon

**Expected behavior**:
Layers can be uploaded to the docker daemon

**`jib-gradle-plugin` Configuration**:
```groovy
jib {
    from.image = "docker://${dockerBaseImageFullPath}"
    to.image = "${dockerTargetImage}"
    container {
      format = 'Docker'
      entrypoint = 'INHERIT'
      ports = ['8080']
      user = 'root'
    }
    extraDirectories {
      paths {
        path {
          from = file("${project.buildDir}/docker")
          into '/opt/app/eldorado-root/'
        }
      }
    }
  }
```

**Log output**: 
```bash
Caused by: com.google.cloud.tools.jib.plugins.common.BuildStepsExecutionException: 'docker load' command failed with error: Error processing tar file(exit status 1): unexpected EOF

	at com.google.cloud.tools.jib.plugins.common.JibBuildRunner.runBuild(JibBuildRunner.java:283)
	at com.google.cloud.tools.jib.gradle.BuildDockerTask.buildDocker(BuildDockerTask.java:115)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at org.gradle.internal.reflect.JavaMethod.invoke(JavaMethod.java:104)
	... 91 more
Caused by: java.io.IOException: 'docker load' command failed with error: Error processing tar file(exit status 1): unexpected EOF

	at com.google.cloud.tools.jib.docker.DockerClient.load(DockerClient.java:211)
	at com.google.cloud.tools.jib.builder.steps.LoadDockerStep.call(LoadDockerStep.java:74)
	at com.google.cloud.tools.jib.builder.steps.StepsRunner.lambda$loadDocker$18(StepsRunner.java:618)
	at com.google.common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleTask.runInterruptibly(TrustedListenableFutureTask.java:125)
	at com.google.common.util.concurrent.InterruptibleTask.run(InterruptibleTask.java:69)
	at com.google.common.util.concurrent.TrustedListenableFutureTask.run(TrustedListenableFutureTask.java:78)
Caused by: java.nio.file.NoSuchFileException: /root/.cache/google-cloud-tools-java/jib/local/8ec1e6e385638003696fd867d91b22c5a03265bafae6c66c37e2ff7bebd4b0f9/f5d031af01f137ae07fa71720fab94d16cc8a2a59868766002918b7c240f3967
	at com.google.cloud.tools.jib.blob.FileBlob.writeTo(FileBlob.java:38)
	at com.google.cloud.tools.jib.tar.TarStreamBuilder.writeAsTarArchiveTo(TarStreamBuilder.java:52)
	at com.google.cloud.tools.jib.image.ImageTarball.dockerWriteTo(ImageTarball.java:149)
	at com.google.cloud.tools.jib.image.ImageTarball.writeTo(ImageTarball.java:78)
	at com.google.cloud.tools.jib.docker.DockerClient.load(DockerClient.java:197)
	... 5 more
	Suppressed: java.io.IOException: This archive contains unclosed entries.
		at org.apache.commons.compress.archivers.tar.TarArchiveOutputStream.finish(TarArchiveOutputStream.java:289)
		at org.apache.commons.compress.archivers.tar.TarArchiveOutputStream.close(TarArchiveOutputStream.java:307)
		at com.google.cloud.tools.jib.tar.TarStreamBuilder.writeAsTarArchiveTo(TarStreamBuilder.java:55)
		at com.google.cloud.tools.jib.image.ImageTarball.dockerWriteTo(ImageTarball.java:149)
		at com.google.cloud.tools.jib.image.ImageTarball.writeTo(ImageTarball.java:78)
		at com.google.cloud.tools.jib.docker.DockerClient.load(DockerClient.java:197)
		at com.google.cloud.tools.jib.builder.steps.LoadDockerStep.call(LoadDockerStep.java:74)
		at com.google.cloud.tools.jib.builder.steps.StepsRunner.lambda$loadDocker$18(StepsRunner.java:618)
		at com.google.common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleTask.runInterruptibly(TrustedListenableFutureTask.java:125)
		at com.google.common.util.concurrent.InterruptibleTask.run(InterruptibleTask.java:69)
		at com.google.common.util.concurrent.TrustedListenableFutureTask.run(TrustedListenableFutureTask.java:78)
		at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
		at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
		at java.base/java.lang.Thread.run(Thread.java:834)
```

**Additional Information**:
Building the image locally without --paralell works fine. Layers can be uploaded to the docker daemon.
The size of the image is 630 MB

Enough disk space inside the pipeline job was available

### PR DESCRIPTION
**Pull request by chanseokoh**
Fixes #2829.

Creates a symbolic link instead of a 0-byte empty file for a symbolic link entry in a tarball when extracting.

The test tarball `symlinks.tar` has two symlinks (one for a directory and the other for a file).

```
.
├── directory1
│   └── file-symlink -> ../directory2/regular
├── directory2
│   └── regular
└── directory-symlink -> directory1
```

### CODE REVIEW

File: jib-core/src/main/java/com/google/cloud/tools/jib/cache/CacheStorageWriter.java [hunk 1]
```diff
@@ -120,8 +120,8 @@ private static void moveIfDoesNotExist(Path source, Path destination) throws IOE
                 () -> {
                   if (Files.exists(destination)) {
                     // If the file already exists, we skip renaming and use the existing file.
-                    // This happens if a new layer happens to have the same content as a
-                    // previously-cached layer.
+                    // This happens, e.g., if a new layer happens to have the same content as a
+                    // previously-cached layer or the same layer is being cached concurrently.
```
Comment (mpeddada1) [inline jib-core/src/main/java/com/google/cloud/tools/jib/cache/CacheStorageWriter.java:124]: Does this mean if the same layer is being cached twice?

File: jib-core/src/main/java/com/google/cloud/tools/jib/tar/TarExtractor.java [hunk 2]
```diff
@@ -61,8 +63,14 @@ public static void extract(Path source, Path destination) throws IOException {
           if (entryPath.getParent() != null) {
             Files.createDirectories(entryPath.getParent());
           }
-          try (OutputStream out = new BufferedOutputStream(Files.newOutputStream(entryPath))) {
-            ByteStreams.copy(tarArchiveInputStream, out);
+
+          String linkName = entry.getLinkName();
```
Comment (loosebazooka) [inline jib-core/src/main/java/com/google/cloud/tools/jib/tar/TarExtractor.java:67]: Can we use `entry.isSymbolicLink`?

File: jib-core/src/main/java/com/google/cloud/tools/jib/tar/TarExtractor.java [hunk 3]
```diff
@@ -61,8 +63,14 @@ public static void extract(Path source, Path destination) throws IOException {
...
+          if (!Strings.isNullOrEmpty(linkName)) {
+            Files.createSymbolicLink(entryPath, Paths.get(linkName));
```
Comment (loosebazooka) [inline jib-core/src/main/java/com/google/cloud/tools/jib/tar/TarExtractor.java:69]: So what happens when we then process this layer? Do we lose this symlink?

**Review by loosebazooka — APPROVED** [review]

File: jib-core/src/main/java/com/google/cloud/tools/jib/cache/CacheStorageWriter.java (hunk 1 above)
Comment (chanseokoh) [reply jib-core/src/main/java/com/google/cloud/tools/jib/cache/CacheStorageWriter.java:124]: Yes. To be strict, multiple threads may be attempting to cache the same layer more than once (whether almost at the same time or apart). They will all write files at their own temporary location. In the end, all of them will try to rename the temp directory to the final one, which is unique. Then, only one rename will succeed, and other will see that the directory already exists.

File: jib-core/src/main/java/com/google/cloud/tools/jib/tar/TarExtractor.java (hunk 3 above)
Comment (chanseokoh) [reply jib-core/src/main/java/com/google/cloud/tools/jib/tar/TarExtractor.java]: The `.extract()` method is only used when using a local tar or Docker daemon base image. And we'll see a symlink only when there are duplicate layers. I learned that when our code reads through the symlink, it actually reads the original file. It's not possible that you read "something" out of the symlink itself.

File: jib-core/src/main/java/com/google/cloud/tools/jib/tar/TarExtractor.java (hunk 2 above)
Comment (chanseokoh) [reply jib-core/src/main/java/com/google/cloud/tools/jib/tar/TarExtractor.java]: Oh, absolutely. Nice catch.

File: jib-core/src/main/java/com/google/cloud/tools/jib/tar/TarExtractor.java (hunk 3 above)
Comment (loosebazooka) [reply jib-core/src/main/java/com/google/cloud/tools/jib/tar/TarExtractor.java]: Ah right, so we only see symlinks to the layer tar (but we never expand out those layer tars).

**Pull request merged**